### PR TITLE
FEATURE: Replace `Crawl-delay` directive with proper rate limiting

### PR DESCRIPTION
**Pull request by OsamaSayegh**
We have a couple of site setting, `slow_down_crawler_user_agents` and `slow_down_crawler_rate`, that are meant to allow site owners to signal to specific crawlers that they're crawling the site too aggressively and that they should slow down.

When a crawler is added to the `slow_down_crawler_user_agents` setting, Discourse currently adds a `Crawl-delay` directive for that crawler in `/robots.txt`. Unfortunately, many crawlers don't support the `Crawl-delay` directive in `/robots.txt` which leaves the site owners no options if a crawler is crawling the site too aggressively.

This PR replaces the `Crawl-delay` directive with proper rate limiting for crawlers added to the `slow_down_crawler_user_agents` list. On every request made by a non-logged in user, Discourse will check the User Agent string and if it contains one of the values of the `slow_down_crawler_user_agents` list, Discourse will only allow 1 request every N seconds for that User Agent (N is the value of the `slow_down_crawler_rate` setting) and the rest of requests made within the same interval will get a 429 response. 

The `slow_down_crawler_user_agents` setting becomes quite dangerous with this PR because if something too generic is added to the setting, it can potentially rate limit most if not all of anonymous traffic. So to protect against this scenario, we've added a new validation to the setting so it errors out when the setting is changed if any of the values are shorter than 3 characters.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -958,4 +959,25 @@ def activated_themes_json
     ids = Theme.transform_ids(id)
     Theme.where(id: ids).pluck(:id, :name).to_h.to_json
   end
+
+  def rate_limit_crawlers
+    return if current_user.present?
+    user_agent = request.user_agent&.downcase
+    return if user_agent.blank?
+
+    SiteSetting.slow_down_crawler_user_agents.downcase.split("|").each do |crawler|
```
Comment (SamSaffron) [inline app/controllers/application_controller.rb:968]: no need to downcase and split a blank, I would just early return prior to user_agent check even if slow_down_crawler_user_agents.blank?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -958,4 +959,25 @@ def activated_themes_json
...
+        key = "#{crawler}_crawler_rate_limit"
+        limiter = RateLimiter.new(
+          nil,
+          key,
+          1,
+          SiteSetting.slow_down_crawler_rate,
+          error_code: key
+        )
+        limiter.performed!
+        return
```
Comment (SamSaffron) [inline app/controllers/application_controller.rb:979]: return from a block is somewhat confusing, I would just break out. 

```
irb(main):001:0> [1,2,3].each {|x| p x; break}
1
=> nil
```

File: lib/site_settings/validations.rb [hunk 3]
```diff
@@ -201,6 +201,13 @@ def validate_cors_origins(new_val)
     validate_error :cors_origins_should_not_have_trailing_slash
   end
 
+  def validate_slow_down_crawler_user_agents(new_val)
+    return if new_val.blank?
+    if new_val.split("|").any? { |crawler| crawler.size < 3 }
+      validate_error :slow_down_crawler_user_agent_must_be_at_least_3_characters
```
Comment (SamSaffron) [inline lib/site_settings/validations.rb:207]: I see about wanting the extra validation here, I guess we can test on  not matching `["gecko","mozilla","firefox","khtml","windows","linux","apple","webkit"]` and a few other obvious strings.

**Comment by SamSaffron** [comment]
overall looks great.

File: lib/site_settings/validations.rb (hunk 3 above)
Comment (OsamaSayegh) [reply lib/site_settings/validations.rb]: done!

**Review by SamSaffron — APPROVED** [review]

**Pull request merged**